### ISSUE TITLE
clEnqueueMemsetINTEL not implemented

### ISSUE THREAD

**Issue by jandrovins** [pocl/pocl#1223]
I am using POCL as SYCL's OpenCL backend. When using memset on a sycl queue, it calls `clEnqueueMemsetINTEL`, which gives the following error when `POCL_DEBUG=all`: 

```
  *** ERROR ***  |     ERROR |  unknown platform extension requested: clEnqueueMemsetINTEL
```
When I search for `clEnqueueMemsetINTEL` in `clGetExtensionFunctionAddressForPlatform.c` it is indeed not there, and I think `clEnqueueMemsetINTEL` only appears on the file `cl_ext.h`:

```
typedef cl_int (CL_API_CALL *
clEnqueueMemsetINTEL_fn)(
    cl_command_queue command_queue,
    void* dst_ptr,
    cl_int value,
    size_t size,
    cl_uint num_events_in_wait_list,
    const cl_event* event_wait_list,
    cl_event* event) ;

#ifndef CL_NO_PROTOTYPES

extern CL_API_ENTRY cl_int CL_API_CALL
clEnqueueMemsetINTEL(
    cl_command_queue command_queue,
    void* dst_ptr,
    cl_int value,
    size_t size,
    cl_uint num_events_in_wait_list,
    const cl_event* event_wait_list,
    cl_event* event) ;

#endif /* CL_NO_PROTOTYPES */

```

From this I understand that this function is not implemented in POCL. Is this by design or is it a bug? Will it be implemented in the future?

**Comment by jansol** [comment]
Our `cl_ext.h` is just the standard Khronos header, it lists symbols for all registered extensions regardless of whether PoCL supports them.

As for whether `clEnqueueMemsetINTEL` will be supported, that depends on whether someone needs the extension it belongs to (`cl_intel_unified_shared_memory`) and decides to implement it or not. So far it seems nobody has. Although even if someone does I'm not sure if that exact function would be implemented since it appears to have been removed from the extension in a newer revision and the header actually has a comment above to the definition saying that you should be using `clEnqueueMemFillINTEL` instead.

**Comment by franz** [comment]
Yep according to [docs](https://github.com/KhronosGroup/OpenCL-Docs/blob/main/extensions/cl_intel_unified_shared_memory.asciidoc), `clEnqueueMemsetINTEL` was removed in 2019 already, SYCL needs an update. OTOH this should be fairly easy to implement in PoCL by calling `clEnqueueMemFillINTEL`.

**Comment by jandrovins** [comment]
I see. I would love to implement it but I currently lack the time. I changed my SYCL code to instead use the `queue.fill()` method.

Thank you both for your help.

**Comment by pjaaskel** [comment]
Is this now from using SYCL on DPC++ on PoCL-CPU on ARM? How well does it work for you in general?

**Comment by jandrovins** [comment]
Yes, this is on an ARM CPU. It works generally well. I have successfully run several applications using POCL as SYCL's backend. I have experienced some problems, but it is to be expected, I think, as the environment gets a bit complicated. For some of these problems I have created issues here and others I have solved on my own. The installation is a bit complicated, but in general terms it works OK. I have only used SYCL programs that use USM, and very few (I can remember only one) that uses the SYCL accessors. But for USM I can say it works OK. I should also say I have not tried programs that use very advanced SYCL functionality... so yeah, in general, it works well
